feat(frontend): Implement test for register hook

stevefaeembra opened the pull request:
Extra tests

stevefaeembra committed:
feat(frontend): Implement test for register hook

Extra tests

marksmall commented on client/src/accounts/authentication/register.hook.test.tsx:
```diff
@@ -0,0 +1,117 @@
+    await waitFor(() => {
+      expect(result.current.isError).toBe(true);
+      expect(result.current.error).toEqual({ message: expectedMessage });
+    });
```
This doesn't follow the pattern I pinned (see last pinned comment) to the dev channel, is there a reason for this? I expected to see

```
await waitFor(() => expect(result.current.isSuccessful).toBe(true);

expect(result.current.isError).toBe(true);
expect(result.current.error).toEqual({ message: expectedMessage });
```

It is a common mistake, according to Kent Dodds to have multiple assertions in a `waitFor`, read this [blog](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#having-multiple-assertions-in-a-single-waitfor-callback). I've linked to the exact section, but please re-read the whole blog, might fill in some gaps in your knowledge.

marksmall commented on client/src/accounts/authentication/register.hook.test.tsx:
```diff
@@ -0,0 +1,117 @@
+    act(() => {
+      result.current.mutate(registerForm);
+    });
+    await waitFor(() => {
```
Please use empty lines to separate out logical blocks of code, it just makes it easier to read, for instance, the act of submitting the form, is logically separate from the verifying the result is as expected.

marksmall commented on client/src/accounts/authentication/register.hook.test.tsx:
```diff
@@ -0,0 +1,117 @@
+  });
+
+  it('should return if registration was successful', async () => {
+    // happy path
```
ths comment helps no one, so remove. What should it return if successful, it does return something doesn't it? If not then  say something like `should return successful response if registration completed`.

marksmall reviewed (changes requested):
couple of tidy ups, to make better use of whitespace to separate logical code blocks. Another request for change to use the same pattern used in testing the `error-report.hook.test.ts`. All similar tests, should work the same way.

stevefaeembra committed:
fix(frontend): Apply PR feedback

Had to reintroduce cross-fetch fetch monkey patch in setup-test
Node's fetch implementation (undici) seems flaky, working one
day and not the next. It is experimental after all, so I think
it's best to stick to known good implementation

marksmall commented on client/setupTests.ts:
```diff
@@ -2,8 +2,17 @@
+// https://github.com/apollographql/apollo-client/issues/10785#issuecomment-1527830658)
+global.fetch = nodeFetch;
+global.Request = Request;
+global.Response = Response;
```
I don't see this as the solution, but there again, you don't explain what the problem is your seeing. I can't say I've had any trouble with fetch, it is no longer experimental in the version of NodeJS we are using either. I do admit, I've not cloned this to run the tests with/without this though, but without an explanation to what it is fixing, I'm dubious of the solution.

The link you refer to is do do with apollo graphql, something we don't use, so why should that be listened too? Having read the comment, it is talking about NextJS undici, again something we are not using, so I don't see any of this as being relevant.

I have had no trouble at all with node and fetch, been using it all weekend writing my own versions of these tests and had no trouble.

Another issue I have with this solution is that we don't directly have `node-fetch` as a dependency, so it is just a quirk of some other dependency that it even exists in our setup, that makes this solution very brittle, if that dependency were updated/removed, this too breaks.

marksmall commented on client/src/accounts/authentication/register.hook.test.tsx:
```diff
@@ -88,12 +84,12 @@ describe('useRegisterHook', () => {
 
-  it('should return if registration was successful', async () => {
-    // happy path
+  it('should show success code if registration was successful', async () => {
```
It doesn't show the success code, it returns it as part of the response, nothing in a hook test is shown.

marksmall commented on client/src/accounts/authentication/register.hook.test.tsx:
```diff
@@ -109,9 +105,11 @@ describe('useRegisterHook', () => {
-      expect(result.current.data).toEqual(1);
     });
+
+    expect(result.current.status).toEqual('success');
```
The correct matcher here is `toBe`, remember, `toBe` for basic types, `int|boolean|string`, `toEqual` for more complex types e.g. `array|object|function`

stevefaeembra committed:
fix(frontend): Fix tests to remove monkey patch

Needed to rejig tests so they test for isError/isSuccess,
rather than testing the inverse (like isError=false)

stevefaeembra committed:
fix(frontend): Fix tests to use toBe/strictEqual

should be toBe() for basic types, toStrictEqual() for
object comparisons. Also, fix for ESLint.

marksmall commented on client/src/accounts/authentication/register.hook.test.tsx:
```diff
@@ -35,7 +39,7 @@ describe('useRegisterHook', () => {
 
     await waitFor(() => expect(result.current.isError).toBe(true));
-    expect(result.current.error).toEqual({ message: expectedMessage });
+    expect(result.current.error).toStrictEqual({ message: expectedMessage });
```
What made you change to this? I had hoped it was `eslint-plugin-jest-dom`, but having undone this change, nothing identified this as to be changed. The point of the jest-dom plugin is to coax us to use the right matcher, having looked into it further, there is another plugin `eslint-plugin-jest`, but we don't use that one. The reason why we use the jest-dom one is because it is referenced in the [testing-library docs](https://testing-library.com/docs/ecosystem-eslint-plugin-jest-dom). I'm not saying this is wrong, just trying to understand how to got to the decision to make this change.

stevefaeembra replied:
sorry, my bad. I'd misread your other comment and had replaced the toEqual() with toBe() for the object. This of course failed - it's not a basic type - but it suggested using toStrictEqual() instead. The tests passed.

Looking at the docs, toStrictEqual() does a deep-comparison like toEqual(), but also checks the type. As I'm not specifying a type for the error message, the two will be identical. I should either (a) keep toStrictEqual but add a type for the return message or (b) revert to toEqual()

marksmall replied:
I'm more curious about the `but it suggested using` comment, what suggested, your IDE intellisense, or eslint?

marksmall replied:
I would like us to get guidance from eslint tools, there are plugins that statically analyse the code and can suggest a better matcher e.g. you use `toEqual` on a basic type and it shows a warning/error stating to change this to `toBe` or vice versa,that kind of thing. If it is the IDE intellisense, then I suspect we are all using different extensions and will get different results, I'd like the advice to be consistent if we can.

marksmall replied:
We have this capability in orbis, I'm just not convinced it is working in our `app-auth` client, even though it is in the eslint rules. I need to do a bit of digging into this, to prove it either way.

stevefaeembra replied:
I think it might be the Testing framework? ESLint didn't trip, I saw no error highlights in the IDE, but I got this suggestion in the test output when I used toBe() on the object.

```
AssertionError: expected { message: 'Please re-authenticate' } to be { message: 'Please re-authenticate' } // Object.is equality

If it should pass with deep equality, replace "toBe" with "toStrictEqual"

Expected: { message: 'Please re-authenticate' }
Received: serializes to the same string
```

marksmall reviewed (commented):
You've made a couple of changes, but none related to the comments made, so try to fix all comments as well. Sadly I've not really got the time to look at lots of small changes, that don't relate to comments made.

I'm curious about the `toStrictEqual`, I'd like the system to inform us what the right matcher to use is and as my comment states, we have one and the reason why we use it, but I don't know if we should also add `eslint-plugin-jest` or replace the dom one, so keen to know what made you change from `toEqual` to `toStrictEqual`.
